include/nbl/builtin/hlsl/emulated/emulated_float64_t_impl.hlsl

```diff
@@ -0,0 +1,327 @@
+}
+
+NBL_CONSTEXPR_INLINE_FUNC uint64_t unpackUint64(uint32_t2 val)
+{
+    return ((uint64_t(val.x) & 0x00000000FFFFFFFFull) << 32) | uint64_t(val.y);
+}
+
+template<bool FlushDenormToZero>
+inline uint64_t castFloat32ToStorageType(float32_t val)
+{
+    if (FlushDenormToZero)
+    {
+        const uint64_t sign = uint64_t(ieee754::extractSign(val)) << 63;
+        if (tgmath::isinf(val))
+            return ieee754::traits<float64_t>::inf | sign;
+        uint32_t asUint = ieee754::impl::bitCastToUintType(val);
+        const int f32BiasedExp = ieee754::extractBiasedExponent(val);
+        if (f32BiasedExp == 0)
+            return sign;
+        const uint64_t biasedExp = uint64_t(f32BiasedExp - ieee754::traits<float32_t>::exponentBias + ieee754::traits<float64_t>::exponentBias) << (ieee754::traits<float64_t>::mantissaBitCnt);
+        const uint64_t mantissa = (uint64_t(ieee754::traits<float32_t>::mantissaMask) & asUint) << (ieee754::traits<float64_t>::mantissaBitCnt - ieee754::traits<float32_t>::mantissaBitCnt);
+
+        return sign | biasedExp | mantissa;
+    }
+    else
+    {
+        // static_assert(false);
+        return 0xdeadbeefbadcaffeull;
+    }
+};
+
+NBL_CONSTEXPR_INLINE_FUNC bool isZero(uint64_t val)
+{
+    return (val << 1) == 0;
+}
+
+inline uint64_t castToUint64WithFloat64BitPattern(uint64_t val)
+{
+    if (isZero(val))
+        return val;
+
+#ifndef __HLSL_VERSION
+    int exp = findMSB(val);
+#else
+    int exp = 63;
+    uint64_t mask = ieee754::traits<float64_t>::signMask;
+    while (!(val & mask))
+    {
+        --exp;
+        mask >>= 1;
+    }
+
+
+    //uint32_t2 valPacked = packUint64(val);
+    //int exp = valPacked.x ? firstbithigh(valPacked.x) + 32 : firstbithigh(valPacked.y);
+    //exp = 63 - exp;
+#endif
+    uint64_t mantissa;
+
+    int shiftCnt = 52 - exp;
+    if (shiftCnt >= 0)
+    {
+        mantissa = val << shiftCnt;
+    }
+    else
+    {
+        const int shiftCntAbs = -shiftCnt;
+        uint64_t roundingBit = 1ull << (shiftCnt - 1);
+        uint64_t stickyBitMask = roundingBit - 1;
+        uint64_t stickyBit = val & stickyBitMask;
+
+        mantissa = val >> shiftCntAbs;
+
+        if ((val & roundingBit) && (!stickyBit))
+        {
+            bool isEven = mantissa & 1;
+            if (!isEven)
+                mantissa++;
+        }
+        else if ((val & roundingBit) && (stickyBit || (mantissa & 1)))
+            val += roundingBit;
+
+        //val += (1ull << (shiftCnt)) - 1;
+        //mantissa = val >> shiftCntAbs;
+
+        if (mantissa & 1ull << 53)
+        {
+            mantissa >>= 1;
+            exp++;
+        }
+    }
+    mantissa &= ieee754::traits<float64_t>::mantissaMask;
+    const uint64_t biasedExp = uint64_t(ieee754::traits<float64_t>::exponentBias + exp) << ieee754::traits<float64_t>::mantissaBitCnt;
+
+    return biasedExp | mantissa;
+};
+
+inline uint64_t castToUint64WithFloat64BitPattern(int64_t val)
```

Comment:
templates and their specializations are "safer" than overloads, because HLSL allows for nasty conversions when calling functions